bpo-45677: Reword first section of `sqlite3` docs (#29326)
python/cpython: Doc/library/sqlite3.rst

compliant with the DB-API 2.0 specification described by :pep:`249`, and
requires SQLite 3.7.15 or newer.

To use the module, start by creating a :class:`Connection` object that
represents the database.  Here the data will be stored in the
:file:`example.db` file::

   import sqlite3
   con = sqlite3.connect('example.db')

> Use the special path name ``:memory:`` to create a temporary database in RAM.

Reviewer: I don't think the imperative mood works here — it makes it sound like something that *must* be done, rather than something which *can* be done. If the goal is to avoid addressing the reader as "you" at all costs, you could use the passive voice here:

```suggestion
The special path name ``:memory:`` can be provided to create a temporary database in RAM.
```

But, I generally try to avoid the passive voice, personally — so I think addressing the reader as "you" might be the least bad option in this instance?

Reviewer 2: I hear "use ... to ..." as short for "use ... in order to ...". This could be made explicit with "in order to", but I think it's clear that it's not a command. If it were a command, I would ask for it to be made explicit with "One must use..." etc

Reviewer: I think it's partly the context of the rest of the documentation, which is written very much like a step-by-step recipe, that makes "use ... to" sound not quite right to me. But, just my opinion :)

Author: Personally, I don't think the existing wording sounds like a command, _but_ I like the suggested wording :) Thanks for helping out, both of you! Highly appreciated.

> Now, create a :class:`Cursor` object and call its :meth:`~Cursor.execute`

Reviewer: ```suggestion
Once a :class:`Connection` has been established, create a :class:`Cursor` object and call its :meth:`~Cursor.execute`
```

   # Just be sure any changes have been committed or they will be lost.
   con.close()

> The saved data is persistent and is available in subsequent sessions::

Reviewer: I'm not satisfied the existing language is clear enough about the meaning of the words "persistent" or "session"

```suggestion
The saved data is persistent: it can be reloaded in a subsequent session even after restarting the Python interpreter::
```

Author: Personally, I think this information is superfluous. Isn't it obvious that data stored in a file on disk is persistent? I didn't remove it in this PR, though, as it is out of scope for this PR.

Reviewer: I think that's a good point. I think my main objection with the original documentation is that the second clause is just a repetition of the same thing, without any clarification: "it is persistent" and "it is available in subsequent sessions" have identical semantics, which would be fine if the second clause was a clarification on the first clause -- but to me, if you don't understand the meaning of the first clause, you're probably not going to understand the meaning of the second clause either. But, perhaps you're right -- it's probably out of scope for this PR, so feel free to resolve this comment :)

.. _sqlite3-placeholders:

SQL operations will usually need to use values from Python variables. It is not

> adviced to assemble queries using Python's string operations, because they are

Reviewer 2: ```suggestion
advised to assemble queries using Python's string operations, because they are
```
But something stronger might be good:
```suggestion
SQL operations will usually need to use values from Python variables. It is poor
practice to assemble queries using Python's string operations, because they are
```

Author: Thanks, @jacobtylerwalls! I've applied your suggestions.

> vulnerable to an SQL injection attacks (see the `xkcd webcomic

Reviewer 2: ```suggestion
vulnerable to SQL injection attacks (see the `xkcd webcomic
```

> SQL operations will usually need to use values from Python variables. It is poor
> practice to assemble queries using Python's string operations, because they are
> vulnerable to SQL injection attacks (see the `xkcd webcomic
> <https://xkcd.com/327/>`_ for a humorous example of what can go wrong)::

Reviewer: I think addressing the reader as "you" is better here, because it is somewhat of an instruction to the reader of what _not_ to do.

```suggestion
SQL operations usually need to use values from Python variables. However, you
should beware of using Python's string operations to assemble queries, as they
are vulnerable to SQL injection attacks (see the `xkcd webcomic
<https://xkcd.com/327/>`_ for a humorous example of what can go wrong)::
```

Author: How about this:

```suggestion
SQL operations usually need to use values from Python variables. However,
beware of using Python's string operations to assemble queries, as they
are vulnerable to SQL injection attacks (see the `xkcd webcomic
<https://xkcd.com/327/>`_ for a humorous example of what can go wrong)::
```

Reviewer: I considered that possibility. I still think it's marginally clearer as a direct instruction to the reader, but I can live with this wording :)

   # Never do this -- insecure!
   symbol = 'RHAT'
   cur.execute("SELECT * FROM stocks WHERE symbol = '%s'" % symbol)

Instead, use the DB-API's parameter substitution by putting placeholders in the

> query string. Provide the actual values as a :class:`tuple` as the second
> argument to the cursor's :meth:`~Cursor.execute` method. An SQL statement may

Reviewer: ```suggestion
Instead, use the DB-API's parameter substitution. To insert a variable into
a query string, use a placeholder in the string, and substitute the actual
values into the query by providing them as a :class:`tuple` of values to
the second argument of the cursor's :meth:`~Cursor.execute` method. An
SQL statement may
```

Author: Let me think a little bit about this.

Author: Applied. I find it easier to read now. Thanks!

Reviewer: @erlend-aasland My pleasure 🙂 (If you fancy returning the favour on my PR for the `GenericAlias` docs, I'd be very grateful... but no worries if not!!)

Author: I'll see if I can have a look tomorrow!
